reporters/ClickhouseReporter.js

```diff
@@ -0,0 +1,217 @@
+                targets: [],
+            };
+        });
+        return this.ready;
+    }
+
+    /**
+     * Called whenever site was crawled (either successfully or not)
+     * 
+     * @param {{site: string, failures: number, successes: number, urls: number, data: import('../crawler').CollectResult, crawlTimes: Array<Array<number>>, fatalError: Error, numberOfCrawlers: number, regionCode: string}} data 
+     */
+    update(data) {
+        this.processSite(data.data);
+    }
+
+    /**
+     * @param {import('../crawler').CollectResult} data
+     */
+    processSite(data) {
+        // @ts-ignore
+        const pageId = createUrlHash(new URL(data.initialUrl));
+        this.ready = this.ready.then(async () => {
+            this.queue.pages.push([this.crawlId, pageId, data.testStarted, data.testFinished, data.initialUrl, data.finalUrl, data.timeout]);
+            if (data.data.requests) {
+                const requestRows = data.data.requests.map((request, requestId) => [
+                    this.crawlId, pageId, requestId, request.url, request.method, request.type, request.status,
+                    request.size, request.remoteIPAddress, JSON.stringify(request.responseHeaders || {}),
+                    request.responseBodyHash, request.failureReason, request.redirectedTo, request.redirectedFrom,
+                    request.initiators.map(u => u.replace(/'/g, '')), request.time || 0
+                ]);
+
+                this.queue.requests = this.queue.requests.concat(requestRows);
+            }
+            if (data.data.elements) {
+                this.queue.elements.push([this.crawlId, pageId, data.data.elements.present, data.data.elements.visible]);
+            }
+            if (data.data.cmps) {
+                const cmpRows = data.data.cmps.map(c => [this.crawlId, pageId, c.name, c.isOpen, c.optOutRuns, c.optOutSucceeds, c.error]);
+                this.queue.cmps = this.queue.cmps.concat(cmpRows);
+            }
+            if (data.data.apis) {
+                const {callStats,savedCalls} = data.data.apis;
+                const callStatRows = Object.keys(callStats).map(source => [this.crawlId, pageId, source, JSON.stringify(callStats[source])]);
+                this.queue.apiCallStats = this.queue.apiCallStats.concat(callStatRows);
+                const savedCallRows = savedCalls.map((c, i) => [this.crawlId, pageId, i, c.source, c.description, c.arguments]);
+                this.queue.apiSavedCalls = this.queue.apiSavedCalls.concat(savedCallRows);
+            }
+            if (data.data.cookies) {
+                this.queue.cookies = this.queue.cookies.concat(data.data.cookies.map((c, i) => [this.crawlId, pageId, i, JSON.stringify(c)]));
+            }
+            if (data.data.targets) {
+                this.queue.targets = this.queue.targets.concat(data.data.targets.map((t, i) => [this.crawlId, pageId, i, t.url, t.type]));
+            }
+
+            if (this.queue.pages.length >= 10) {
```

Comment:
what's that? what queue?

Reply:
This is to batch inserts 10 pages at a time.